skrub/_datetime_encoder.py

```diff
@@ -18,78 +23,473 @@
+
+    indices : list of int, default=None
+        Indices of the parsable columns to convert.
+        If None, indices are computed using the current input X.
+
+    index_to_format : mapping of int to str, default=None
+        Dictionary mapping column indices to their datetime format.
+        It defines the format parameter for each column when calling
+        pd.to_datetime.
+
+        If indices is None, ``indices_to_format`` is computed using the
+        current input X.
+        If format is not None, all values of ``indices_to_format`` are set
+        to format.
+
+    format : str, default=None
+        When format is not None, it overwrites the values in indices_to_format.
+
+    Returns
+    -------
+    X_split : list of 1d array of length n_features
+    """
+    if indices is None:
+        indices, index_to_format = _get_datetime_column_indices(X_split)
+
+    # format overwrite indices_to_format
+    if format is not None:
+        index_to_format = {col_idx: format for col_idx in indices}
+
+    for col_idx in indices:
+        X_split[col_idx] = pd.to_datetime(
+            X_split[col_idx], format=index_to_format[col_idx], **kwargs
+        )
+
+    return X_split
+
+
+def _get_datetime_column_indices(X_split, dayfirst=True):
+    """Select the datetime parsable columns by their indices \
+    and return their datetime format.
+
+    Parameters
+    ----------
+    X_split : list of 1d array of length n_features
+
+    Returns
+    -------
+    datetime_indices : list of int
+        List of parsable column, identified by their indices.
+
+    index_to_format: mapping of int to str
+        Dictionary mapping parsable column indices to their datetime format.
+    """
+    indices = []
+    index_to_format = {}
+
+    for col_idx, X_col in enumerate(X_split):
+        X_col = X_col[pd.notnull(X_col)]
+
+        # convert pd.TimeStamp to np.datetime64
+        if all(isinstance(val, pd.Timestamp) for val in X_col):
+            X_col = X_col.astype("datetime64")
+
+        if _is_column_datetime_parsable(X_col):
+            indices.append(col_idx)
+            # TODO: pass require_dayfirst to _guess_datetime_format
+            index_to_format[col_idx] = _guess_datetime_format(X_col)
+
+    return indices, index_to_format
+
+
+def _is_column_datetime_parsable(X_col):
+    """Check whether a 1d array can be converted into a \
+    :class:`pandas.DatetimeIndex`.
+
+    Parameters
+    ----------
+    X_col : array-like of shape ``(n_samples,)``
+
+    Returns
+    -------
+    is_dt_parsable : bool
+    """
+    # Remove columns of int, float or bool casted as object.
+    # Pandas < 2.0.0 raise a deprecation warning instead of an error.
+    with (
+        warnings.catch_warnings()
+        if not _is_pandas_format_mixed_available()
+        else nullcontext()
+    ):
+        if not _is_pandas_format_mixed_available():
+            warnings.simplefilter("ignore", category=DeprecationWarning)
```

Comment:
I guess we could just filter warnings for all pandas versions

Reply:
Yes, but this might be informative for Pandas > 2.0.0 as we are not expecting this warning, WDYT? The logic is not pleasant to read, though.

Reply:
I'm wondering if it's worth it just to avoid masking a potential different DeprecationWarning raised by astype with a recent pandas. but it's not a big deal either way

Reply:
Okay, let's apply it for all pandas versions